### PR TITLE
feat: Implement `useTable` state hook

### PR DESCRIPTION
**Pull request by ling1726**
Implments a `useTable` state hook to support selection and sort.

Initially I tried this with [@tanstack/react-table](https://www.npmjs.com/package/react-table), but that library is incredibly generic and relies heavily on render functions which I would like to avoid (for now). That library is also enormous because it supports features that are currently not asked for in Fluent UI

image

Addresses #24226

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<details>
<summary>Unchanged fixtures</summary>

| Package & Exports                                                                                                                                                             |          Size (minified/GZIP) |
| ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------: |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/MultipleComponents.fixture.js'>react-components: Accordion, Button, FluentProvider, Image, Menu, Popover</abbr> | `188.604 kB`<br />`51.784 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ProviderAndTheme.fixture.js'>react-components: FluentProvider & webLightTheme</abbr>                            |  `32.688 kB`<br />`10.736 kB` |

</details>
<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/5d985fac3569513a2808a9f3b93ad73907325e55'>5d985fac3569513a2808a9f3b93ad73907325e55</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)
<p>No significant results to display.</p><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24329/perf-test-react-components/Avatar-mount_base.html">1273</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24329/perf-test-react-components/Avatar-mount.html">1251</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24329/perf-test-react-components/Button-mount_base.html">953</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24329/perf-test-react-components/Button-mount.html">972</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24329/perf-test-react-components/FluentProvider-mount_base.html">1574</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24329/perf-test-react-components/FluentProvider-mount.html">1559</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24329/perf-test-react-components/FluentProviderWithTheme-mount_base.html">638</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24329/perf-test-react-components/FluentProviderWithTheme-mount.html">613</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24329/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">595</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24329/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">596</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24329/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">633</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24329/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">640</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24329/perf-test-react-components/MakeStyles-mount_base.html">1910</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24329/perf-test-react-components/MakeStyles-mount.html">1875</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24329/perf-test-react-components/SpinButton-mount_base.html">2418</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24329/perf-test-react-components/SpinButton-mount.html">2455</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

**Comment by Hotell** [comment]
55kB of JS... wow. good call

File: scripts/tasks/screener.ts [hunk 1]
```diff
@@ -37,6 +35,8 @@ export async function screener() {
     if (!affectedPackages.has(affectedPackageInfo.packageJson.name)) {
       await cancelScreenerRun(screenerConfig, 'skipped');
     } else {
+      const screenerStates = await getScreenerStates(screenerConfig);
```
Comment (Hotell) [inline scripts/tasks/screener.ts:38]: bad rebase? why is this needed for table ?

Comment (ling1726) [reply scripts/tasks/screener.ts:38]: whoops, no idea how it got there

File: packages/react-components/react-table/src/hooks/useMultipleSelection.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  const allRowIds = React.useMemo(() => new Set<RowId>(items.map((item, i) => getRowId(item, i))), [items, getRowId]);
+
+  const toggleSelectAllRows: SelectionStateInternal['toggleSelectAllRows'] = useEventCallback(() => {
+    setSelected(s => {
+      if (s.size === items.length) {
+        return new Set<RowId>();
+      }
+
+      return new Set<RowId>(items.map((item, i) => getRowId(item, i)));
+    });
```
Comment (bsunderhus) [inline packages/react-components/react-table/src/hooks/useMultipleSelection.ts:16]:
```suggestion
    setSelected(current =>  (
      current.size === items.length ? new Set<RowId>() : new Set<RowId>(items.map((item, i) => getRowId(item, i)))
    ));
```

File: packages/react-components/react-table/src/hooks/useSelection.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import { GetRowIdInternal, SelectionStateInternal, UseTableOptions } from './types';
+import { useMultipleSelection } from './useMultipleSelection';
+import { useSingleSelection } from './useSingleSelection';
+
+export function useSelection<TItem>(
```
Comment (bsunderhus) [inline packages/react-components/react-table/src/hooks/useSelection.ts:5]: IMHO, wouldn't this hook be a bad idea? I mean, ideally it would be better if the user directly called `useMultipleSelection` or `useSingleSelection`.

File: packages/react-components/react-table/src/stories/Table/Sort.stories.tsx [hunk 4]
```diff
@@ -79,80 +78,54 @@ const items: Item[] = [
...
-              onClick={onHeaderCellClick(columnKey)}
-              sortDirection={sortColumn === columnKey ? sortDirection : undefined}
-            >
-              {columns[columnKey].label}
-            </TableHeaderCell>
-          ))}
+          <TableHeaderCell {...headerSortProps('file')}>File</TableHeaderCell>
+          <TableHeaderCell {...headerSortProps('author')}>Author</TableHeaderCell>
+          <TableHeaderCell {...headerSortProps('lastUpdated')}>Last updated</TableHeaderCell>
+          <TableHeaderCell {...headerSortProps('lastUpdate')}>Last update</TableHeaderCell>
```
Comment (bsunderhus) [inline packages/react-components/react-table/src/stories/Table/Sort.stories.tsx:121]: Now that looks nice 😎 👏🏻

File: packages/react-components/react-table/src/hooks/useSort.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+  return {
+    sortColumn,
+    sortDirection,
+    sort,
+    setColumnSort,
+    toggleColumnSort,
+    headerSortProps: (columnId: ColumnId) => ({
+      sortDirection: sortColumn === columnId ? sortDirection : undefined,
+      onClick: () => toggleColumnSort(columnId),
```
Comment (bsunderhus) [inline packages/react-components/react-table/src/hooks/useSort.ts:49]: This looks wrong to me, as this callback should be somehow memoized

**Review by bsunderhus — APPROVED** [review]
Looks amazing to me! I'm totally in favour of delegating complex behaviours to hooks!

File: packages/react-components/react-table/src/hooks/useSort.ts (hunk 5 above)
Comment (ling1726) [reply packages/react-components/react-table/src/hooks/useSort.ts:49]: This is impossible, it's similar to creating factory functions for radio `onChange` handlers

File: packages/react-components/react-table/src/hooks/useSelection.ts (hunk 3 above)
Comment (ling1726) [reply packages/react-components/react-table/src/hooks/useSelection.ts:5]: yeah, I don't like it now either, it's a part of this issue to decompose these
